extensions/src/main/java/org/pentaho/platform/plugin/services/security/userrole/oauth/PentahoOAuthAzureHandler.java

```diff
@@ -0,0 +1,221 @@
+
+package org.pentaho.platform.plugin.services.security.userrole.oauth;
+
+import org.apache.commons.collections.MapUtils;
+import org.apache.commons.lang.StringUtils;
+import org.pentaho.platform.api.engine.security.IAuthenticationRoleMapper;
+import org.pentaho.platform.api.engine.security.userroledao.IUserRoleDao;
+import org.pentaho.platform.api.mt.ITenant;
+import org.pentaho.platform.security.userroledao.PentahoOAuthUser;
+import org.pentaho.platform.util.oauth.PentahoOAuthProperties;
+import org.pentaho.platform.util.oauth.PentahoOAuthUtility;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class PentahoOAuthAzureHandler implements IPentahoOAuthHandler {
+
+  private static final String VALUE = "value";
+
+  IUserRoleDao userRoleDao;
+
+  RestTemplate restTemplate;
+
+  IAuthenticationRoleMapper oauthRoleMapper;
+
+  PentahoOAuthProperties pentahoOAuthProperties;
+
+  Map<String, String> registrationToClientCredentialsToken = new HashMap<>();
+
+  Map<String, Map<String, String>> registrationToAppRoles = new HashMap<>();
+
+
+  public PentahoOAuthAzureHandler( IUserRoleDao userRoleDao,
+                                   RestTemplate restTemplate,
+                                   IAuthenticationRoleMapper oauthRoleMapper,
+                                   PentahoOAuthProperties pentahoOAuthProperties ) {
+    this.userRoleDao = userRoleDao;
+    this.restTemplate = restTemplate;
+    this.oauthRoleMapper = oauthRoleMapper;
+    this.pentahoOAuthProperties = pentahoOAuthProperties;
+  }
+
+  public String getClientCredentialsToken( String registrationId, boolean renewToken ) {
+    String clientCredentialsToken = registrationToClientCredentialsToken.get( registrationId );
+
+    if ( renewToken || StringUtils.isBlank( clientCredentialsToken ) ) {
+      HttpHeaders headers = new HttpHeaders();
+      headers.setContentType( MediaType.APPLICATION_FORM_URLENCODED );
+
+      MultiValueMap<String, String> map = new LinkedMultiValueMap<>();
+      map.add( "client_id", pentahoOAuthProperties.getValue( registrationId + ".client-id" ) );
+      map.add( "client_secret", pentahoOAuthProperties.getValue( registrationId + ".client-secret" ) );
+      map.add( "grant_type", "client_credentials" );
+      map.add( "redirect_uri", pentahoOAuthProperties.getValue( registrationId + ".redirect-uri" ) );
+      map.add( "scope", "https://graph.microsoft.com/.default" );
+
+      HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>( map, headers );
+
+      String url = pentahoOAuthProperties.getValue( registrationId + ".token-uri" );
+      var responseEntity = restTemplate.postForEntity( url, request, Map.class );
```

Comment:
We need to wrap this with error handling and timeout check.

Reply:
https://github.com/pentaho/pentaho-platform/pull/5879/files#diff-40c79d4770a8476ca6b3d538c06e76b85a51603ad4d9cb70d12b3f18cb13367bR160-R165

https://github.com/pentaho/pentaho-platform/pull/5879/files#diff-845d5c754b6456ae8b13b48c5de382ea197a9ff3543cd30ec622a565f6a2664bR66-R67

Added these 2 snippets to handle the same